include/onnxruntime/core/framework/stream_handles.h

```diff
@@ -0,0 +1,159 @@
+    for (const auto& kv : clock) {
+      auto it = other_stream_clock_.find(kv.first);
+      if (it == other_stream_clock_.end()) {
+        other_stream_clock_.insert(kv);
+      } else {
+        other_stream_clock_[kv.first] = std::max(it->second, kv.second);
+      }
+    }
+  }
+
+ protected:
+  StreamHandle handle_;
+  const OrtDevice& device_;
+  uint64_t timestamp_{0};
+  // TODO: use inline container.
+  // currently this class is header only, but abseil doesn't compile with nvcc
+  // we need to add new symbol to provider_bridge and hide abseil from the header.
+  std::unordered_map<Stream*, uint64_t> other_stream_clock_{};
+};
+
+namespace synchronize {
+// an object which record the status of the stream, and can be wait on from another stream.
+class Notification {
+ public:
+  explicit Notification(Stream* s) : stream_(s) {}
+  virtual ~Notification() = default;
+
+  // this api will perform three operations:
+  // 1. activate the notification on device, for example, record an event on GPU.
+  // 2. take a snapshot of the timestamp lookup table in current stream.
+  // 3. bump the timestamp for current stream.
+  void ActivateAndUpdate() {
+    Activate();
+    stream_->CloneCurrentStreamSyncTable(stream_clock_);
+    stream_clock_[stream_] = stream_->BumpTimeStampAndReturn();
+  }
+
+  // return the timestamp lookup table saved in the notificaiton.
+  const std::unordered_map<Stream*, uint64_t>& GetStreamSyncTable() {
+    return stream_clock_;
+  }
+
+ protected:
+  virtual void Activate() = 0;
+  // which stream create this notification.
+  Stream* stream_;
+  // TODO: use inline container.
+  // currently this class is header only, but abseil doesn't compile with nvcc
+  // we need to add new symbol to provider_bridge and hide abseil from the header.
+  std::unordered_map<Stream*, uint64_t> stream_clock_{};
+};
+}  // namespace synchronize
+
+// the definition for the handle for stream commands
+// EP can register the handle to the executor.
+// in the POC, just use primitive function pointer
+// TODO: use a better way to dispatch handles.
+using WaitNotificationFn = std::function<void(Stream&, synchronize::Notification&)>;
+using CreateStreamFn = std::function<std::unique_ptr<Stream>(const OrtDevice&)>;
+
+// an interface of a simple registry which hold the handles EP registered.
+// make it interface so we can pass it through shared library based execution providers
+class IStreamCommandHandleRegistry {
+ public:
+  virtual ~IStreamCommandHandleRegistry() = default;
+  // Wait is a little special as we need to consider the source stream the notification generated, and the stream we are waiting.
+  // i.e., for an cuda event what notify the memory copy, it could be wait on a CPU stream, or on another cuda stream.
+  [[nodiscard]] virtual WaitNotificationFn GetWaitHandle(OrtDevice::DeviceType notification_ower_device_type, OrtDevice::DeviceType executor_device_type) const = 0;
```

Comment:
```suggestion
  [[nodiscard]] virtual WaitNotificationFn GetWaitHandle(OrtDevice::DeviceType notification_owner_device_type,
                                                         OrtDevice::DeviceType executor_device_type) const = 0;
```

also in general, 120 character line length